### PR TITLE
Jump to line feature

### PR DESCRIPTION
**Pull request by helfi92**
To experiment with this feature, select a line number in the log and then refresh the page. the log will take you to the selected line.

### CODE REVIEW

File: src/LogViewer/index.js [hunk 1]
```diff
@@ -23,6 +23,8 @@ export default class LogViewer extends React.Component {
       showLineNumbers: qs.showLineNumbers !== 'false',
       isLoading: true,
       chunkHeights: [],
+      offset: 0,
+      LINE_CHUNK: 0,
```
Comment (eliperelman) [inline src/LogViewer/index.js]: This property should not be uppercased.

File: src/LogViewer/index.js [hunk 2]
```diff
@@ -59,7 +61,20 @@ export default class LogViewer extends React.Component {
   }
 
   componentDidMount() {
-    this.request();
+    const { highlightStart } = this.state;
```
Comment (eliperelman) [inline src/LogViewer/index.js]: `highlightStart` and the jump line should not be related for this log viewer. I know GitHub works differently, but we should introduce a new property for a `jumpToLine` in `LogViewer` so we can move to lines separately from highlighting them.

File: src/LogViewer/chunk-worker.js [hunk 3]
```diff
@@ -115,7 +115,7 @@ const update = (response) => {
     lineCounts = lineCounts.slice(overage);
   }
 
-  self.postMessage(JSON.stringify({ type: 'update', chunkHeights }));
+  self.postMessage(JSON.stringify({ type: 'update', chunkHeights, offset, LINE_CHUNK }));
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: `LINE_CHUNK` is a constant to this file, but being something passed to `LogViewer` makes it not constant to the other file, so it should not be uppercased here. Something like:

``` js
{  lineChunk: LINE_CHUNK  }
```

File: src/LazyList/index.js [hunk 4]
```diff
@@ -15,7 +15,8 @@ export default class LazyList extends React.Component {
     super(props);
 
     this.state = {
-      renderIndices: Range(0, 0)
```
Comment (eliperelman) [inline src/LazyList/index.js:18]: The functionality seems OK, but having the it located in `LazyList` would violate this component's concerns. `LazyList` should only care about the loading and unloading of list items based on scroll position. It does not have knowledge of lines, or the contents of its lists, or how to jump to a certain line. This functionality should probably be in the `LogViewer` component, by using the line knowledge to change the scroll position, which would then cause `LazyList` to update its loaded containers, which should already be handled.

**Review by eliperelman — CHANGES_REQUESTED** [review]

**Comment by helfi92** [comment]
@eliperelman, I totally agree with you. Keeping the components functional is the way to go. I can implement a `jumpToLine` function in LogViewer that can take a `lineNumber` and scroll the window by an x amount of pixels and then let LazyList take control. I am aware there is currently no UI that can take a `lineNumber`. Do we need one or is that not important for now.
I can also implement a `jumpToHighlight` function which will run if and only if there is a `highlightStart` number in the URL. I think that's a great feature to have. Thoughts?

**Comment by helfi92** [comment]
@eliperelman latest commit addresses your comments. Looks a lot cleaner :)

File: src/LogViewer/chunk-worker.js [hunk 5]
```diff
@@ -115,7 +115,10 @@ const update = (response) => {
     lineCounts = lineCounts.slice(overage);
   }
 
-  self.postMessage(JSON.stringify({ type: 'update', chunkHeights }));
+  self.postMessage(JSON.stringify({
+    type: 'update', chunkHeights, offset,
```
Comment (eliperelman) [inline src/LogViewer/chunk-worker.js]: If you multi-line object properties, put each one on their own line, e.g.:

``` js
self.postMessage(JSON.stringify({
  chunkHeights,
  offset,
  type: 'update',
  lineChunk: LINE_CHUNK,
  minLineHeight: MIN_LINE_HEIGHT
}));
```

File: src/LogViewer/index.js (hunk 2 above)
Comment (eliperelman) [reply src/LogViewer/index.js]: I still think this `highlightStart` stuff needs to be removed.

File: src/LogViewer/index.js [hunk 6]
```diff
@@ -178,12 +183,29 @@ export default class LogViewer extends React.Component {
 
   renderChunks() {
     return (
-      <LazyList list={this.state.chunkHeights} buffer={2}>
+      <LazyList
```
Comment (eliperelman) [inline src/LogViewer/index.js]: This can stay on 1 line.

File: src/LogViewer/index.js [hunk 7]
```diff
@@ -178,12 +182,29 @@ export default class LogViewer extends React.Component {
...
+        list={this.state.chunkHeights}
+        buffer={2}>
         {(heights, shouldLoad) => heights.map((height, index) => this.renderChunk(height, index, shouldLoad))}
       </LazyList>
     );
   }
 
+  jumpToLine(lineNumber) {
+    const { minLineHeight, offset } = this.state;
```
Comment (eliperelman) [inline src/LogViewer/index.js:193]: I don't see this method being called outside of `jumpToHighlight`. I think we need to fetch the `lineNumber` for jumping from a new query string/state entry, and then handle jumping once it's loaded, similar to how it's being done in `jumpToHighlight`.

**Review by eliperelman — CHANGES_REQUESTED** [review]

**Comment by helfi92** [comment]
Ok, this should do what you suggested. Requesting a review.

**Review by eliperelman — COMMENTED** [review]
Code looks good, but the line jumping didn't work precisely for me. I tried the following URL:

http://localhost:4000/?url=https%3A%2F%2Fpublic-artifacts.taskcluster.net%2F5qB6Zqc0Tw2_wkKwHgGNGQ%2F0%2Fpublic%2Flogs%2Flive_backing.log&lineNumber=3000

and it put me around line 2550.

**Comment by helfi92** [comment]
@eliperelman weird, but for this specific log, the line after 1446 turns into 1001. 

screen shot 2016-10-13 at 6 44 00 pm

Works on the huge log though:
http://localhost:4000/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FB8yyH1HORxKkQ9p1Sh0CFA%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive_backing.log&highlightStart=&highlightEnd=&lineNumber=300000

Thoughts?

**Comment by eliperelman** [comment]
@helfi92 I'm not sure. You may want to try adding some `console.log`s to the code to see if you can debug what the expected coordinates, percentages, and window calculations are.

**Comment by eliperelman** [comment]
Looks like the line differences are because of some HTML non-escaping in the log. Will fix that in a follow up.

**Pull request merged**